CORE-2866 Fix assignable object states

uskudnik opened the pull request:
We must ensure:
- only verifiers can verify requests
- editing Request/attaching evidence/URL on Open or Verified/Finished  moves request back to "In Progress" (with a warning on any of such actions beforehand)
- mapping objects should not move request back to `In Progress`

See jira ticket for details.

jurev commented on src/ggrc/assets/javascripts/mustache_helper.js:
```diff
@@ -2342,6 +2342,40 @@ Mustache.registerHelper("if_auditor", function (instance, options) {
   return options.inverse(options.contexts);
 });
 
+Mustache.registerHelper("if_verifier", function (instance, options) {
```
Since this renders something, wouldn't a better name for this be render_verifier or sth.?

uskudnik replied:
It doesn't render something in a conventional sense of printing out something, `defer_render` (I presume that's why you think of that) is used here just because we have to wait resolving block inside the tag until `refresh_instances` returns and only after it returns we decide whether something is allowed or not - there is no nice way for using Deferred here.

jurev replied:
Aha OK, thanks for clarification.

plamut commented on src/ggrc/migrations/versions/20151203153139_297131e22e28_add_final_state_to_request_status_and_.py:
```diff
@@ -0,0 +1,27 @@
+  op.execute("""ALTER TABLE requests CHANGE status status ENUM("Open","In Progress","Finished","Verified","Final","Unstarted") NOT NULL;""")
+  op.execute("""UPDATE requests SET status="Unstarted" WHERE status="Open";""")
+  op.execute("""UPDATE requests SET status="Finished" WHERE status="Final";""")
+  op.execute("""ALTER TABLE requests CHANGE status status ENUM("Unstarted","In Progress","Finished","Verified") NOT NULL;""")
```
Seeing that we use multi line string syntax, should we perhaps split longer SQL statements into multiple lines for better readability?

uskudnik replied:
Usually I would go with yes but in this case I don't think it's a major concern/issue and would leave it as it is (I have some bad experiences with `op.execute` and multiple statements with it and I would not push it since it's of minimal added value for general code quality).

zidarsk8 replied:
@uskudnik Actually execute has issues with multiple statements and not multi line. And I do agree with @plamut here.

But If you don't want to split it up, replace the """ with " and add ```# noqa``` in the end.

plamut commented on src/ggrc_gdrive_integration/assets/javascripts/controllers/gdrive_workflows_controller.js:
```diff
@@ -1111,76 +1111,90 @@ can.Component.extend({
+      };
+
+      $.when(verify_dfd).then(dfd.then(function() {
+        //dfd.then(function () {
```
Should be removed if not needed anymore?

uskudnik replied:
Negative or rather, not yet - this is WIP and this particular segment is waiting for @hyperNURb .

plamut replied:
OK, good, makes sense.

uskudnik replied:
We should completely re-do this entire segmet :) If you take a look this commit is marked as WIP and will most probably be deleted altogether from history when Ivan makes PR and I'll probably rebase it.

plamut replied:
Haha, true that about rewriting. :)
Yes, I am aware of the WIP label, thus all feedback should be considered intermediary and by no means final, no worries. :)

Smotko commented:
Confirm button should be green and I believe we don't really need a cancel button, since we already have a little cross in the top right.

![screen shot 2015-12-16 at 12 37 30](https://cloud.githubusercontent.com/assets/513444/11839900/d7d34a88-a3f1-11e5-8845-39c286e7e8b5.png)

uskudnik commented:
@Smotko Replaced confirmation buttons, fixed `trigger_upload_parent` bug that I found not refreshing instance, added `finish` button to open request, fixed bug that showed warning on creating new Request

Smotko commented:
I found another issue, if the state is OPEN we probably don't want to show the dialog to move it back in progress:

![screen shot 2015-12-17 at 16 09 33](https://cloud.githubusercontent.com/assets/513444/11873001/ac45a7d4-a4d8-11e5-87d8-a68fb8b55790.png)

uskudnik commented:
@Smotko Fixed the wording, left the dialog since this is about to trigger change.

Smotko commented on src/ggrc/assets/mustache/base_templates/dates_list.mustache:
```diff
@@ -21,9 +21,11 @@
           <input tabindex="5" class="span8 input-block-level date" name="due_on" data-toggle="datepicker" data-after="created_at" placeholder="MM/DD/YYYY" type="text" value="{{localize_date due_on}}">
         {{else}}
           {{#is_allowed 'update' instance}}
```
this is_allowed should be `{{#is_allowed 'update' instance context='for'}}`

Smotko replied:
I'll create a separate PR for this.

Smotko commented on src/ggrc/models/mixins.py:
```diff
@@ -669,10 +669,10 @@ class Assignable(object):
                  for r in self.related_sources
                  if "AssigneeType" in r.attrs]
-    assignees += [(r.destination, r.attrs["AssigneeType"].split(","))
+    assignees += [(r.destination, tuple(r.attrs["AssigneeType"].split(",")))
```
What was the reason for adding a tuple here? @edofic do you agree with this change?

uskudnik replied:
`.split` outputs lists which aren't hashable (when converting to sets).

edofic replied:
:+1:

Smotko commented:
It seems ok, but the amount of code changes are making me a bit nervous. @uskudnik please do a dry run of these new features tomorrow on grc-test.
